csharp/src/Google.Protobuf/CodedInputReader.cs

```diff
@@ -0,0 +1,816 @@
+            int startField = WireFormat.GetTagFieldNumber(startGroupTag);
+            int endField = WireFormat.GetTagFieldNumber(tag);
+            if (startField != endField)
+            {
+                throw new InvalidProtocolBufferException(
+                    $"Mismatched end-group tag. Started with field {startField}; ended with field {endField}");
+            }
+            recursionDepth--;
+        }
+
+        /// <summary>
+        /// Reads a double field from the stream.
+        /// </summary>
+        public double ReadDouble()
+        {
+            return BitConverter.Int64BitsToDouble((long)ReadRawLittleEndian64());
+        }
+
+        /// <summary>
+        /// Reads a float field from the stream.
+        /// </summary>
+        public float ReadFloat()
+        {
+            const int length = sizeof(float);
+
+            if (reader.CurrentSpan.Length - reader.CurrentSpanIndex >= length)
+            {
+                // Fast path. All data is in the current span.
+                ReadOnlySpan<byte> data = reader.CurrentSpan.Slice(reader.CurrentSpanIndex, length);
+                reader.Advance(length);
+
+                return Unsafe.ReadUnaligned<float>(ref MemoryMarshal.GetReference(data));
+            }
+            else
+            {
+                return ReadFloatSlow();
+            }
+        }
+
+        private unsafe float ReadFloatSlow()
+        {
+            const int length = sizeof(float);
+
+            byte* buffer = stackalloc byte[length];
+            Span<byte> tempSpan = new Span<byte>(buffer, length);
+
+            ThrowEndOfInputUnless(reader.TryCopyTo(tempSpan));
+            reader.Advance(length);
+
+            return Unsafe.ReadUnaligned<float>(ref MemoryMarshal.GetReference(tempSpan));
+        }
+
+        /// <summary>
+        /// Reads a uint64 field from the stream.
+        /// </summary>
+        public ulong ReadUInt64()
+        {
+            return ReadRawVarint64();
+        }
+
+        /// <summary>
+        /// Reads an int64 field from the stream.
+        /// </summary>
+        public long ReadInt64()
+        {
+            return (long)ReadRawVarint64();
+        }
+
+        /// <summary>
+        /// Reads an int32 field from the stream.
+        /// </summary>
+        public int ReadInt32()
+        {
+            return (int)ReadRawVarint32();
+        }
+
+        /// <summary>
+        /// Reads a fixed64 field from the stream.
+        /// </summary>
+        public ulong ReadFixed64()
+        {
+            return ReadRawLittleEndian64();
+        }
+
+        /// <summary>
+        /// Reads a fixed32 field from the stream.
+        /// </summary>
+        public uint ReadFixed32()
+        {
+            return ReadRawLittleEndian32();
+        }
+
+        /// <summary>
+        /// Reads a bool field from the stream.
+        /// </summary>
+        public bool ReadBool()
+        {
+            return ReadRawVarint32() != 0;
```

Comment:
should be ReadRawVarint64() != as was discovered here: https://github.com/protocolbuffers/protobuf/pull/6910

Reply:
Same for ReadBoolWrapper (if applicable).